compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -856,11 +864,63 @@ impl<'hir> LoweringContext<'_, 'hir> {
             // Lower outside new scope to preserve `is_in_loop_condition`.
             let fn_decl = this.lower_fn_decl(decl, None, FnDeclKind::Closure, None);
 
+            if let &ClosureBinder::For { span: for_sp, .. } = binder {
+                fn span_of_infer(ty: &hir::Ty<'_>) -> Option<Span> {
+                    use intravisit::Visitor;
+
+                    struct V(Option<Span>);
+
+                    impl<'v> Visitor<'v> for V {
+                        fn visit_ty(&mut self, t: &'v hir::Ty<'v>) {
+                            match t.kind {
+                                _ if self.0.is_some() => (),
+                                hir::TyKind::Infer => {
+                                    self.0 = Some(t.span);
+                                }
+                                _ => intravisit::walk_ty(self, t),
+                            }
+                        }
+                    }
+
+                    let mut v = V(None);
+                    v.visit_ty(ty);
+                    v.0
+                }
+
+                for input in fn_decl.inputs {
+                    if let Some(sp) = span_of_infer(input) {
+                        self.sess
+                            .struct_span_err(
+                                sp,
+                                "implicit parameter types are forbidden when `for<...>` is present",
+                            )
+                            .span_label(for_sp, "`for<...>` is here")
+                            .emit();
```

Comment:
Could you issue a single diagnostic for all instances of `TyKind::Infer`?

Reply:
Done in 34c87a68f93f33d663c7a4b60d9708abf5c8e35f